server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/VacationRoutes.java

```diff
@@ -0,0 +1,199 @@
+
+    private final JsonTransformer jsonTransformer;
+    private final JsonExtractor<VacationDTO> jsonExtractor;
+
+    private final VacationRepository vacationRepository;
+    private final NotificationRegistry notificationRegistry;
+    private final UsersRepository usersRepository;
+
+    @Inject
+    public VacationRoutes(VacationRepository vacationRepository, NotificationRegistry notificationRegistry,
+                          UsersRepository usersRepository, JsonTransformer jsonTransformer) {
+        this.vacationRepository = vacationRepository;
+        this.notificationRegistry = notificationRegistry;
+        this.usersRepository = usersRepository;
+        this.jsonTransformer = jsonTransformer;
+        this.jsonExtractor = new JsonExtractor<>(VacationDTO.class, new JavaTimeModule());
+    }
+
+    @Override
+    public String getBasePath() {
+        return VACATION;
+    }
+
+    @Override
+    public void define(Service service) {
+        service.get(VACATION + SEPARATOR + USER_NAME, this::getVacation, jsonTransformer);
+        service.post(VACATION + SEPARATOR + USER_NAME, this::updateVacation);
+        service.patch(VACATION + SEPARATOR + USER_NAME, this::setVacation);
+        service.delete(VACATION + SEPARATOR + USER_NAME, this::deleteVacation);
+    }
+
+    public VacationDTO getVacation(Request request, Response response) {
+        testUserExists(request);
+        try {
+            AccountId accountId = AccountId.fromString(request.params(USER_NAME));
+            Vacation vacation = vacationRepository.retrieveVacation(accountId).block();
+            return VacationDTO.from(vacation);
+        } catch (IllegalStateException e) {
+            LOGGER.info("Invalid get on user vacation", e);
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.NOT_FOUND_404)
+                .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                .message("Invalid get on user vacation")
+                .cause(e)
+                .haltError();
+        }
+    }
+
+    public String updateVacation(Request request, Response response) {
+        testUserExists(request);
+        try {
+            AccountId accountId = AccountId.fromString(request.params(USER_NAME));
+            VacationDTO vacationDto = jsonExtractor.parse(request.body());
+            VacationPatch vacationPatch = VacationPatch.builder()
+                .subject(updateTo(vacationDto.getSubject()))
+                .textBody(updateTo(vacationDto.getTextBody()))
+                .htmlBody(updateTo(vacationDto.getHtmlBody()))
+                .fromDate(updateTo(vacationDto.getFromDate()))
+                .toDate(updateTo(vacationDto.getToDate()))
+                .isEnabled(updateTo(vacationDto.getEnabled()))
+                .build();
+            vacationRepository.modifyVacation(accountId, vacationPatch)
+                .then(notificationRegistry.flush(accountId)).block();
```

Comment:
A nice refactoring would be to have the `vacationRepository` flushing by itself the `notificationRegistry` (or an abstraction on top of them) so that we don't duplicate that statement...

Reply:
I actually missed that one in my first implementation, so I wholeheartedly agree. If I am refactoring all this anyway, might as well do it right. Will create a fourth commit for that though to keep things separate.